### PR TITLE
feat: Multichain search DB filling

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/10652

## Motivation

Add support of filling the data multi-chain search DB.

### Export explorer chain data to the [Multichain Search DB](https://github.com/blockscout/blockscout-rs/tree/main/multichain-search)

Integration with the service is enabled, if  `MICROSERVICE_MULTICHAIN_SEARCH_URL` env variable contains valid URL.

Export is built-in to synchronous/asynchronous parts of realtime/catchup indexer: everywhere, where we import addresses, blocks or transactions, additional step is added which transforms and exports that data to the Multichain Search service DB via single API endpoint provided by the service.

Multichain search service API endpoint is set via `MICROSERVICE_MULTICHAIN_SEARCH_URL` and `MICROSERVICE_MULTICHAIN_SEARCH_API_KEY` env variables.

Backfilling migration has been added as well, which moves from the head to the tail, fetches in batches (batch size is managed by `MIGRATION_BACKFILL_MULTICHAIN_SEARCH_BATCH_SIZE`), exports existing in the DB addresses, block hashes, transaction hashes and block ranges based on the data from `blocks`, `token transfers`, `transactions` and `internal_transctions` tables.

_Note_: Backfilling migration block ranges may intersect with the block ranges of the catchup indexer. Intersections in the imported data are handled by the Multichain service DB.

### Adding corresponding backfilling migration status API endpoint

GET `/api/health/multichain-search-export` is added to track the status of backfilling of Multichain Search DB.

Response in the case migration is in progress:
```
{"migration":{"finished":false,"metadata":{"max_block_number":6684354}}}
```

Response in the case migration finished:
```
{"migration":{"finished":true,"metadata": null}}
```

In both scenarios 200 status code is returned.

### Chore
- `api/v1/health` endpoints duplicated at `api/health` path. In the future, we'll disable `api/v1/health` endpoints. **The response structure on the health endpoints is changed to improve readability.**

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables). https://github.com/blockscout/docs/pull/338
- [x] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [x] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Release Notes

- **New Features**
  - Introduced health check endpoints: `/health`, `/liveness`, `/readiness`, and `/multichain-search-export`.
  - Added functionality for batch importing data into the Multichain search database.
  - Implemented a new background migration for backfilling the Multichain search database.

- **Bug Fixes**
  - Improved error handling for various API responses, now including structured metadata.

- **Configuration**
  - New configuration options for Multichain search services and backfill operations added, including API key, service URL, concurrency, and batch size.
  - Added environment variables for multichain search configurations.

- **Tests**
  - Updated tests to reflect changes in API response structures and enhanced coverage for edge cases.

- **Documentation**
  - Updated module documentation to clarify expanded capabilities related to data migrations.
  - Enhanced clarity in error handling documentation.

- **Chores**
  - Added "multichain" to the spell check vocabulary.
  - Streamlined environment configuration files by removing unnecessary comments.

### CODE REVIEW

File: config/runtime.exs [hunk 1]
```diff
@@ -556,6 +556,11 @@ config :explorer, Explorer.MicroserviceInterfaces.Metadata,
 config :explorer, Explorer.SmartContract.StylusVerifierInterface,
   service_url: ConfigHelper.parse_microservice_url("MICROSERVICE_STYLUS_VERIFIER_URL")
 
+config :explorer, Explorer.MicroserviceInterfaces.MultichainSearch,
+  api_key: System.get_env("MICROSERVICE_MULTICHAIN_SEARCH_API_KEY"),
+  service_url: System.get_env("MICROSERVICE_MULTICHAIN_SEARCH_URL"),
+  enabled: ConfigHelper.parse_bool_env_var("MICROSERVICE_MULTICHAIN_SEARCH_ENABLED")
```
Comment (sl1depengwyn) [inline config/runtime.exs:562]: I understand that this comes from microservice template use, but maybe we can omit this variable and decide whether the service enabled via `MICROSERVICE_MULTICHAIN_SEARCH_URL`? I think this would be enough and we came up with a similar decision about stylus verifier internally

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    ## Returns
+    - `{:ok, :service_disabled}`: If the integration with Multichain Search Service is disabled.
+    - `{:ok, result}`: If the import was successful.
+    - `{:error, reason}`: If an error occurred.
+  """
+  @spec batch_import(%{
+          addresses: list(),
+          blocks: list(),
+          transactions: list()
+        }) :: {:error, :disabled | <<_::416>> | Jason.DecodeError.t()} | {:ok, any()}
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:35]:
```suggestion
  @spec batch_import(%{
          addresses: [Address.t()],
          blocks: [Block.t()],
          transactions: [Transaction.t()]
        }) :: {:error, :disabled | String.t() | Jason.DecodeError.t()} | {:ok, any()}
```

Also, need to replace `alias Explorer.Chain.Hash` with `alias Explorer.Chain.{Address, Block, Hash, Transaction}`

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+          ]
+
+        transactions_query = from(transaction in Transaction, where: transaction.block_number in ^block_numbers)
+
+        transactions_task =
+          Task.async(fn ->
+            transactions_query
+            |> preload(^transaction_preloads)
+            |> Repo.all(timeout: :infinity)
+          end)
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex:87]: maybe bring this closer to token transfer query

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            # credo:disable-for-next-line Credo.Check.Refactor.Nesting
+            case MultichainSearch.batch_import(to_import) do
+              {:ok, _} = result ->
+                result
+
+              {:error, _} ->
+                Logger.error(fn ->
+                  ["#{@failed_to_export_data_error}", "#{@for}", "#{inspect(block_numbers)}"]
+                end)
+
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex:160]: in batch_import we expect that token field is tried to preload, though here we are not preload token, so all addresses will be formatted as non tokens.
I suggest to remove preloads from the previous request entirely as they are redundant as far as I can see, and preload necessary smart contract and token fields here after whole addresses list, this way we reduce number of requests to the database

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    update_batch(block_numbers)
+  end
+
+  @impl FillingMigration
+  def update_cache do
+    BackgroundMigrations.set_backfill_multichain_search_db_finished(true)
+  end
+
+  @spec extract_address_from_result(Transaction.t() | InternalTransaction.t() | TokenTransfer.t() | Block.t()) :: list()
+  defp extract_address_from_result(result) do
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex:191]: what you think about proxy implementations here? they must be created in a some transaction, so they would be imported as created_contract_address here on other block, but what if they are predeployed? I think we need to add call to multichain search import to chainspec parser

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 6]
```diff
@@ -281,6 +282,12 @@ defmodule Indexer.Block.Fetcher do
       update_uncles_cache(inserted[:block_second_degree_relations])
       update_withdrawals_cache(inserted[:withdrawals])
 
+      update_multichain_search_db(%{
+        addresses: inserted[:addresses],
+        blocks: inserted[:blocks],
+        transactions: inserted[:transactions]
+      })
```
Comment (sl1depengwyn) [inline apps/indexer/lib/indexer/block/fetcher.ex:289]: here we have no any preloads I think

File: apps/explorer/lib/explorer/chain/import.ex [hunk 7]
```diff
@@ -69,7 +69,7 @@ defmodule Explorer.Chain.Import do
 
   @type all_result ::
           {:ok, %{unquote_splicing(quoted_runner_imported)}}
-          | {:error, [Changeset.t()] | :timeout}
+          | {:error, [Changeset.t()] | :timeout | :insert_to_multichain_search_db_failed}
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/import.ex:72]: where exactly this error may come from, as far as I see actual import is not affected with microservice communication

**Review by sl1depengwyn — CHANGES_REQUESTED** [review]
about preloads, I suggest to not preload anything at all before calling MultichainSearch module and preload entites that we expect there, this way we do not need to keep in mind all places where import called, but we can be sure that associations were preloaded

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex (hunk 4 above)
Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex:160]: correction: not entirely, but leave only address preloads itself
maybe even better option is to pass to microservice interface address hashes instead if addresses, however I'm not sure, but this way worth considering

File: apps/explorer/lib/explorer/chain/import.ex (hunk 7 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/import.ex:72]: the spec of `import` function defines output as of `Import.all_result()` type `@spec import(Import.all_options()) :: Import.all_result()`. And one of the possible result of function execution is `_ -> {:error, :insert_to_multichain_search_db_failed}` now.

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+    update_batch(block_numbers)
+  end
+
+  @impl FillingMigration
+  def update_cache do
+    BackgroundMigrations.set_backfill_multichain_search_db_finished(true)
+  end
+
+  @spec extract_address_from_result(Transaction.t() | InternalTransaction.t() | TokenTransfer.t() | Block.t()) :: list()
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex]: Processed in https://github.com/blockscout/blockscout/pull/11139/commits/781976d50da5368629401221370e4ed5d028a308

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v1/health_controller.ex [hunk 9]
```diff
@@ -45,49 +107,104 @@ defmodule BlockScoutWeb.API.V1.HealthController do
...
+  ## Examples
+
+      iex> conn = %Plug.Conn{}
+      iex> multichain_search_db_export(conn, %{})
+      %Plug.Conn{status: 200, resp_body: "{\"migration\":{\"finished\":false,\"meta\":{\"max_block_number\":6684354}}}"}
+  """
+  @spec multichain_search_db_export(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def multichain_search_db_export(conn, _) do
+    migration = MigrationStatus.fetch(@backfill_multichain_search_db_migration_name)
+    migration_status_finished? = migration.status == "completed"
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v1/health_controller.ex]: Indeed, I missed that. Without handling in the case of disabled integration `/api/health/multichain-search-export` will throw an exception. I processed in https://github.com/blockscout/blockscout/pull/11139/commits/7269a3c8ad4bc91f858b0427485c43b68283f1eb.

File: config/runtime.exs (hunk 1 above)
Comment (vbaranov) [reply config/runtime.exs]: Processed in https://github.com/blockscout/blockscout/pull/11139/commits/a00a51ccf79962ed06a2034c469238010625cb0b. PR description and related PR to the docs repo are updated accordingly.

**Comment by vbaranov** [comment]
> about preloads, I suggest to not preload anything at all before calling MultichainSearch module and preload entites that we expect there, this way we do not need to keep in mind all places where import called, but we can be sure that associations were preloaded

@sl1depengwyn I processed the suggested preloads refactoring as we discussed internally. For proxy implementations, I created an enhancement request for the future https://github.com/blockscout/blockscout/issues/11491.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v1/health_controller.ex [hunk 10]
```diff
@@ -45,49 +107,104 @@ defmodule BlockScoutWeb.API.V1.HealthController do
...
+          finished: migration_status_finished?,
+          metadata: migration.meta
+        }
+      }
+      |> Jason.encode!()
+
+    send_resp(conn, :ok, response)
+  end
+
+  defp result(
```
[the remark this answers is not in the document]
Comment (sl1depengwyn) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v1/health_controller.ex:149]: Yes, but maybe even better to leave old format in v1 endpoint during one release, not sure if it worth it though

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+  defp get_token_name(%NotLoaded{}), do: nil
+
+  defp get_token_name(token), do: token.name
+
+  defp get_smart_contract_name(nil), do: nil
+
+  defp get_smart_contract_name(%NotLoaded{}), do: nil
+
+  defp get_smart_contract_name(token), do: token.name
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:156]: maybe rename argument to smart_contract

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+         addresses: raw_addresses,
+         blocks: blocks,
+         transactions: transactions
+       }) do
+    chain_id = NetVersion.get_version()
+    block_ranges = get_block_ranges(blocks)
+
+    addresses =
+      raw_addresses
+      |> Repo.preload([:token, :names, :smart_contract])
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:96]: as far as I can see we do not access names field, so we do not need names preload here

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        transactions_query = from(transaction in Transaction, where: transaction.block_number in ^block_numbers)
+
+        transactions_task =
+          Task.async(fn ->
+            transactions_query
+            |> preload(^transaction_preloads)
+            |> Repo.all(timeout: :infinity)
+          end)
+
+        token_transfer_preloads = [:from_address, :to_address, :token_contract_address]
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex:80]: maybe move this definition closer to `token_transfers_query` definition on line 94?

**Review by sl1depengwyn — APPROVED** [review]

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex (hunk 11 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex]: Processed in https://github.com/blockscout/blockscout/pull/11139/commits/47b77a354dc82b21c99f01de09be651a15a86f13

File: apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex (hunk 13 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/migrator/backfill_multichain_search_db.ex]: Processed in https://github.com/blockscout/blockscout/pull/11139/commits/47b77a354dc82b21c99f01de09be651a15a86f13

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex (hunk 12 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex]: Processed in https://github.com/blockscout/blockscout/pull/11139/commits/47b77a354dc82b21c99f01de09be651a15a86f13
